src/Core/src/Handlers/HybridWebView/HybridWebViewHandler.Android.cs

```diff
@@ -0,0 +1,219 @@
+			}
+
+			public void SendRawMessage(string rawMessage)
+			{
+#pragma warning disable CA1416 // Validate platform compatibility
+				PostWebMessage(new WebMessage(rawMessage), AndroidAppOriginUri);
+#pragma warning restore CA1416 // Validate platform compatibility
+			}
+		}
+
+		public static void MapSendRawMessage(IHybridWebViewHandler handler, IHybridWebView hybridWebView, object? arg)
+		{
+			if (arg is not string rawMessage || handler.PlatformView is not IHybridPlatformWebView hybridPlatformWebView)
+			{
+				return;
+			}
+
+			hybridPlatformWebView.SendRawMessage(rawMessage);
+		}
+	}
+
+	public class AndroidHybridWebViewClient : WebViewClient
+	{
+		private readonly HybridWebViewHandler _handler;
+
+		public AndroidHybridWebViewClient(HybridWebViewHandler handler)
+		{
+			_handler = handler;
+		}
+
+		public override WebResourceResponse? ShouldInterceptRequest(AWebView? view, IWebResourceRequest? request)
+		{
+			var fullUrl = request?.Url?.ToString();
+			var requestUri = HybridWebViewQueryStringHelper.RemovePossibleQueryString(fullUrl);
+
+			if (new Uri(requestUri) is Uri uri && HybridWebViewHandler.AppOriginUri.IsBaseOf(uri))
+			{
+				var relativePath = HybridWebViewHandler.AppOriginUri.MakeRelativeUri(uri).ToString().Replace('/', '\\');
+
+				string contentType;
+				if (string.IsNullOrEmpty(relativePath))
+				{
+					relativePath = _handler.DefaultFile;
+					contentType = "text/html";
+				}
+				else
+				{
+					if (!HybridWebViewHandler.ContentTypeProvider.TryGetContentType(relativePath, out contentType!))
+					{
+						// TODO: Log this
+						contentType = "text/plain";
+					}
+				}
+
+				var assetPath = Path.Combine(_handler.HybridRoot!, relativePath!);
+				var contentStream = PlatformOpenAppPackageFile(assetPath);
+
+				if (contentStream is null)
+				{
+					var notFoundContent = "Resource not found (404)";
+
+					var notFoundByteArray = Encoding.UTF8.GetBytes(notFoundContent);
+					var notFoundContentStream = new MemoryStream(notFoundByteArray);
+
+					return new WebResourceResponse("text/plain", "UTF-8", 404, "Not Found", GetHeaders("text/plain"), notFoundContentStream);
+				}
+				else
+				{
+					// TODO: We don't know the content length because Android doesn't tell us. Seems to work without it!
+					return new WebResourceResponse(contentType, "UTF-8", 200, "OK", GetHeaders(contentType), contentStream);
+				}
+			}
+			else
+			{
+				return base.ShouldInterceptRequest(view, request);
+			}
+		}
+
+		private Stream? PlatformOpenAppPackageFile(string filename)
+		{
+			filename = PathUtils.NormalizePath(filename);
+
+			try
+			{
+				return _handler.Context.Assets?.Open(filename);
+			}
+			catch (Java.IO.FileNotFoundException)
+			{
+				return null;
+			}
+		}
+
+		internal static class PathUtils
+		{
+			public static string NormalizePath(string filename) =>
+				filename
+					.Replace('\\', Path.DirectorySeparatorChar)
+					.Replace('/', Path.DirectorySeparatorChar);
+		}
```

Comment:
The MAUI FileSystem APIs are unfortunately async-only (normally a good thing), but the Android API being implemented here is sync-only. For that reason I did the call directly because ultimately the low-level file system API on Android is sync anyway (just regular file system call). I have been unable to find a pattern to do async interception of Android WebView requests.

Reply:
Can we create an issue to track this, we already have `FileSystemUtils.NormalizePath` so we probably should create a helper for this and use it everywhere vs copying in all places.

Reply:
Filed https://github.com/dotnet/maui/issues/23668.